pkg/controller/apimanagerbackup/apimanagerbackup_logic_reconciler.go

```diff
@@ -0,0 +1,398 @@
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupCompletion()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	return result, err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupInS3Destination() (reconcile.Result, error) {
+	// TODO implement
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupInPVCDestination() (reconcile.Result, error) {
+	var res reconcile.Result
+	var err error
+
+	err = r.reconcileBackupDestinationPVC()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	res, err = r.reconcileBackupDestinationPVCStatus()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileBackupSecretsAndConfigMapsToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileAPIManagerCustomResourceBackupToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileBackupSystemFileStoragePVCToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	return res, err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupDestinationPVC() error {
+	desired := r.apiManagerBackup.BackupDestinationPVC()
+	if desired == nil {
+		return nil
+	}
+
+	// TODO create mutator function for PVC ?
+	err := r.ReconcileResource(&v1.PersistentVolumeClaim{}, desired, reconcilers.CreateOnlyMutator)
+
+	return err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupSecretsAndConfigMapsToPVCJob() (reconcile.Result, error) {
+	desired := r.apiManagerBackup.BackupSecretsAndConfigMapsToPVCJob()
+	if desired == nil {
+		return reconcile.Result{}, nil
+	}
+
+	// TODO create Mutator function for Jobs??
+	existing := &batchv1.Job{}
+	// Check if this step has finished
+	if r.cr.SecretsAndConfigMapsBackupStepFinished() {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if backup substep has completed
+	if !r.cr.SecretsAndConfigMapsBackupSubStepFinished() {
+		err := r.ReconcileResource(existing, desired, reconcilers.CreateOnlyMutator)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		substepFinished := true
+		if existing.Status.Succeeded != *desired.Spec.Completions {
+			r.Logger().Info("Job has still not finished", "Job Name", desired.Name, "Actively running Pods", existing.Status.Active, "Failed pods", existing.Status.Failed)
+			return reconcile.Result{Requeue: true}, nil
+		}
+
+		r.cr.Status.SecretsAndConfigMapsBackupSubStepFinished = &substepFinished
+		err = r.UpdateResourceStatus(r.cr)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		r.Logger().Info("Job finished successfully. Requeing", "Job Name", desired.Name)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Check if cleanup substep has completed
+	if !r.cr.SecretsAndConfigMapsCleanupSubStepFinished() {
+		common.TagToObjectDeleteWithPropagationPolicy(desired, metav1.DeletePropagationForeground)
```

Comment:
I am curious. Why do you need [Foreground cascading deletion](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion). Which are resources owned by the job that needs to be deleted first?

Reply:
I see, created pods are dependent of the job. But the need of foreground cascading deletion I still do not see.

Reply:
You are correct about pods are dependent on the job and they are not deleted automatically unless you provide a cascading deletion policy.

Reply:
Regarding foreground or background cascading deletion, the reason I used foreground is to be sure the Jobs and its dependent elements (Pods) are deleted before going to the next step.
If I were to use background I think it might happen that the Job is deleted but its pods would not be deleted yet. The code checks the jobs (and not its pods) so there might be a scenario where job would be deleted and pods not? (and PVCs are at pod level at the end, and are needed on next stages).